### PR TITLE
feat: add error handling to deserialization

### PR DESCRIPTION
**Pull request by hansieodendaal**
- Fixed VarInt decoding
- Added error handling to deserialization
- Added dynamic memory limits for decoding serialized data
- Fixed subtract with overflow
- Added guards to vector sizing
- Fixed panic 'Custom { kind: Interrupted, error: ScriptNotSupported }'
- Added error handling to ExtraField deserialization
- Added error handling to SubField::Padding() deserialization
- Added fuzzing unit tests
- Added fuzzing proper

Co-Authored-By: stringhandler <https://github.com/stringhandler>

### CODE REVIEW

**Comment by Boog900** [comment]
Thanks for adding fuzzing to the lib!

For the extra field, the padding sub-field does not encode a length, here is the code in Monero: https://github.com/monero-project/monero/blob/90294f09ae34ef96f3dea5fea544816786df87c8/src/cryptonote_basic/tx_extra.h#L55
So the error you got is expected behavior, padding must be the last field or the padding must be the maximum length (255 bytes).

I am also unsure of the benefits of exposing the error on serialization, the trait is sealed, I may be missing something?

> Fixed panic 'Custom { kind: Interrupted, error: ScriptNotSupported }'

I would rather these types be removed from monero-rs, they have never been used in Monero and there was discussion recently about removing them from Monero's code-base

**Comment by hansieodendaal** [comment]
Hi @Boog900, thanks for reviewing. I have added a lot of error handling, basically replacing all `unwrap()` in the code, as well as fixing critical errors I found.

I will deal with the padding issue tomorrow.

**Comment by hansieodendaal** [comment]
@Boog900 this is ready for review now

File: src/consensus/encode.rs [hunk 1]
```diff
@@ -522,21 +619,52 @@ impl<T: Encodable> Encodable for Box<[T]> {
...
     }
 }
 
+// Maximum allocatable memory bytes size - this number does not need to be exact, it only needs to be ample, as its
+// primary purpose is to prevent memory overflows during consensus decoding for vectors or arrays.
+pub(crate) fn max_allocatable_mem_bytes_size(bytes_upper_limit: usize) -> usize {
+    const MAX_MEM_ALLOC_SIZE: usize = 32 * 1024 * 1024; // 32 MiB
+    min(MAX_MEM_ALLOC_SIZE, max(bytes_upper_limit, 128))
+}
+
```
Comment (Boog900) [inline src/consensus/encode.rs:658]: Continuing the other comment if `MAX_MEM_ALLOC_SIZE` is the max then why not just always use the max?

File: src/cryptonote/hash.rs [hunk 2]
```diff
@@ -99,17 +104,25 @@ impl crate::consensus::encode::Encodable for Hash {
...
+    #[error("Encode error: {0}")]
+    EncodeError(#[from] EncodeError),
+}
+
 /// Capacity of an object to hash itself and return the result as a plain [`struct@Hash`] or as an
 /// interpreted scalar value into [`PrivateKey`].
 pub trait Hashable {
     /// Return its own hash.
-    fn hash(&self) -> Hash;
+    fn hash(&self) -> Result<Hash, HashError>;
```
Comment (Boog900) [inline src/cryptonote/hash.rs:119]: This does complicate the API on something that "shouldn't" fail? maybe im missing something again?

File: src/cryptonote/onetime_key.rs [hunk 3]
```diff
@@ -138,29 +141,47 @@ pub struct SubKeyChecker<'a> {
...
-        });
-        SubKeyChecker { table, keys }
+                match subaddress::get_spend_public_key(keys, index) {
+                    Ok(spend) => {
+                        table.insert(spend, index);
+                    }
+                    Err(e) => {
+                        return Err(e);
+                    }
+                }
```
Comment (Boog900) [inline src/cryptonote/onetime_key.rs:164]: cant we just use a `?` here

**Review by Boog900 — CHANGES_REQUESTED** [review]
Ok so I had look over and I don't think I agree with most of the changes.

I don't feel that exposing the error on serialization is necessary and doing so requires us to expose it on hashing as well which just complicates the API on something that shouldn't  panic.

Although I agree `unwrap` and `expect` are usually not great, if you know something *can't* fail then there is no point complicating the API and confusing users.

I don't know about having a bytes_upper_limit on every call to `consensus_decode` especially if we aren't also keeping track of already allocated bytes.

File: src/blockdata/transaction.rs [hunk 4]
```diff
@@ -845,52 +864,72 @@ impl crate::consensus::encode::Encodable for ExtraField {
 }
 
 impl Decodable for SubField {
-    fn consensus_decode<R: io::Read + ?Sized>(r: &mut R) -> Result<SubField, encode::Error> {
-        let tag: u8 = Decodable::consensus_decode(r)?;
+    fn consensus_decode<R: io::Read + ?Sized + Seek>(
```
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/blockdata/transaction.rs:867]: Yip, I should have deleted it, a remnant of my previous padding implementation, thanks.

File: src/cryptonote/onetime_key.rs (hunk 3 above)
Comment (hansieodendaal) [reply src/cryptonote/onetime_key.rs:164]: The previous `for_each` closures did not allow the different match arms, and now we need a closure that handles both the success and failure modes.

**Comment by hansieodendaal** [comment]
Hi @Boog900, thank you for the review. Let me address the comments in a couple of commits and different PRs. The first commit here is to address the comments that I have resolved above. I will also branch out in different PR's where I demonstrate the failure modes.

File: src/cryptonote/hash.rs (hunk 2 above)
Comment (hansieodendaal) [reply src/cryptonote/hash.rs:115]: Please see unit test `fn previous_fuzz_transaction_hash_failures()` in #182

File: src/consensus/encode.rs [hunk 5]
```diff
@@ -26,76 +26,98 @@
...
+    /// A consensus encoding parsing error.
+    #[error("Parsing error: {0}")]
+    ConsensusEncodingFailed(String),
 }
 
 /// Encode an object into a vector of byte.
-pub fn serialize<T: Encodable + std::fmt::Debug + ?Sized>(data: &T) -> Vec<u8> {
+pub fn serialize<T: Encodable + std::fmt::Debug + ?Sized>(
+    data: &T,
+) -> Result<Vec<u8>, EncodeError> {
```
[the remark this answers is not in the document]
Comment (hansieodendaal) [reply src/consensus/encode.rs:74]: Please see unit test `fn previous_fuzz_extra_field_try_parse_failures()` in #182

**Comment by hansieodendaal** [comment]
Closing this PR in favour of some smaller PRs

**Pull request closed without merging**